### PR TITLE
Address comments on Gzip implementation

### PR DESCRIPTION
**Pull request by schneems**
- don't mutate PATH_INFO in env, test
- test fallback content type matches Rack::File
- change assertion style
- make HTTP_ACCEPT_ENCODING comparison case insensitive
- return gzip path from method instead of true/false so we don't have to assume later
- don't allocate un-needed hash.

Original comments:

https://github.com/rails/rails/commit/cfaaacd9763642e91761de54c90669a88d772e5a

cc @jeremy

### CODE REVIEW

**Comment by chrisnicola** [comment]
@schneems will this eliminate the need for middlewares like:

https://github.com/mattolson/heroku_rails_deflate
or
https://github.com/romanbsd/heroku-deflater

Both of those also enable `Rack::Deflater` to compress application requests as well as assets, assuming they are not already compressed.

I'm guessing with these changes, and assuming pre-compiled compressed static assets are available, one would put `Rack::Deflater` after `ActionDispatch::Static` rather than before it and there should be no need for either of the Heroku specific middlewares. Am I correct?

File: guides/source/4_2_release_notes.md [hunk 1]
```diff
@@ -268,6 +268,11 @@ Please refer to the [Changelog][action-pack] for detailed changes.
 *   Added an option to disable logging of CSRF failures.
     ([Pull Request](https://github.com/rails/rails/pull/14280))
 
+*   When the Rails server is set to serve static assets, gzip assets will now be
+    served if the client supports it and a pre-generated gzip file (.gz) is on disk.
```
Comment (chrisnicola) [inline guides/source/4_2_release_notes.md]: Is Sprockets generating them by default again? I don't see that in any of the commits https://github.com/rails/sprockets. Also the discussion left with the suggestion that compressions would be opt-in not the default. https://github.com/rails/sprockets/issues/26

**Comment by chancancode** [comment]
@schneems sorry I missed your response, a brief note in the upgrade guide seems good!

**Comment by schneems** [comment]
> @schneems will this eliminate the need for middlewares like

It's better to serve an already compressed image/asset rather than having to read the whole thing in from disk and then compress on the fly. I would honestly benchmark the performance impact, my gut instinct would be to say to put them after Static, but I don't know if it would save a meaningful amount of time per request without trying it out. Maybe those gems know if a request is already compressed?

> a brief note in the upgrade guide seems good!

If generating gzipped files is removed from sprockets (which it currently is), then this change doesn't actually do anything other than add an extra file system call to every asset request :frowning: If we don't get that behavior turned back on in sprockets we should consider removing this behavior. Alternatively if sprockets enables it based on a flag we could use the same flag in AD::Static to enable/disable the check.

**Comment by chrisnicola** [comment]
@schneems yes, I believe that's exactly what those two middlewares do. They compress on the fly using `Rack::Deflater` only if they don't find a static asset first. If they do they just serve the compressed asset.

I'd suggest that they are not really necessary assuming Sprockets can compress assets and ActionStatic can serve them up by default. So I think it would make sense now to just recommend just putting `Rack::Deflater` after `ActionDispatch::Static` generally, unless you really need to compress assets on the fly.

**Pull request merged**